File: linked_lists/questions/question_2.2/problem_2.2.rb
require_relative './../../../custom_data_structures/ruby_data_structures/linked_list.rb'

class LinkedList

  def kth_to_last(k)
    leader = advance_head_k_elements(k)
    follower = @head
    return nil unless leader
    while leader.next
      leader = leader.next
      follower = follower.next
    end
    follower.value
  end

  def advance_head_k_elements(k)
    runner = @head
    k.times do
      return nil unless runner.next
      runner = runner.next
    end
    runner
  end
end

File: linked_lists/questions/question_2.2/problem_2.2.js
LinkedList.prototype.kthToLast = function(k){
  // debugger
  if (this.isEmpty()){
    return null;
  }
  var runner = this.advanceRunner(k);
  var kBehind = this.head
  if (!runner){
    return null;
  }
  while(runner.next){
    runner = runner.next;
    kBehind = kBehind.next;
  }
  return kBehind;
}

LinkedList.prototype.advanceRunner = function(n){
  var count = 0;
  var runner = this.head
  while (count < n){
    runner = runner.next;
    count++
    if (!runner){
      return null;
    }
  }
  return runner;
}
